src/participant/prompts/history.ts

```diff
@@ -2,36 +2,129 @@ import * as vscode from 'vscode';
+        continue;
+      }
+
       for (const fragment of historyItem.response) {
-        if (
-          fragment instanceof vscode.ChatResponseMarkdownPart &&
-          historyItem.result.metadata?.responseContent
-        ) {
-          res += fragment.value.value;
+        if (fragment instanceof vscode.ChatResponseMarkdownPart) {
+          message += fragment.value.value;
+
+          if (historyItem.result.metadata?.askForNamespace) {
+            // When the message is the assistant asking for part of a namespace,
+            // we only want to include the question asked, not the user's
+            // database and collection names in the history item.
+            break;
+          }
         }
       }
       // eslint-disable-next-line new-cap
-      messages.push(vscode.LanguageModelChatMessage.Assistant(res));
+      messages.push(vscode.LanguageModelChatMessage.Assistant(message));
     }
-  });
+  }
 
   return messages;
 }
+
+// A user could mistakenly write one of these which would mess up our parsing.
+// To fix this we either need to pass metadata with the command, which
+// is doesn't look like it possible, or we need to have a more complex
+// message key, like using a special symbol for this regex (although a user could
+// copy paste this themselves and get into another weird state).
+export const databaseNameInCommandResponseKey = 'Database: ';
+const databaseNameInCommandResponseKeyRegex = `^${databaseNameInCommandResponseKey}(.*)`;
+export const collectionNameInCommandResponseKey = 'Collection: ';
+const collectionNameInCommandResponseKeyRegex = `^${collectionNameInCommandResponseKey}(.*)`;
+
+export function getLatestDatabaseAndCollectionFromChatHistory({
+  request,
+  context,
+}: {
+  request: vscode.ChatRequest;
+  context: vscode.ChatContext;
+}): {
+  databaseName: string | undefined;
+  collectionName: string | undefined;
+} {
+  let databaseName: string | undefined;
+  let collectionName: string | undefined;
+
+  context.history.forEach(
+    (historyItem: vscode.ChatRequestTurn | vscode.ChatResponseTurn, index) => {
+      if (historyItem instanceof vscode.ChatResponseTurn) {
+        const possibleNamespaceResponse =
+          (context.history[index - 1] as vscode.ChatRequestTurn).prompt || '';
+        const databaseNameMatch = possibleNamespaceResponse.match(
+          databaseNameInCommandResponseKeyRegex
+        )?.[1];
+        databaseName = databaseNameMatch || databaseName;
+        const collectionNameMatch = possibleNamespaceResponse.match(
+          collectionNameInCommandResponseKeyRegex
+        )?.[1];
+        collectionName = collectionNameMatch || collectionName;
+
+        if (historyItem.result.metadata?.databaseName) {
+          databaseName = historyItem.result.metadata?.databaseName;
+        }
+        if (historyItem.result.metadata?.collectionName) {
+          collectionName = historyItem.result.metadata?.collectionName;
+        }
```

Comment:
The order of operations here seems incorrect as well. There's no reason to first try and parse the namespace from the prompt if we're going to then replace it with the namespace in the response metadata. It also relies on alternating request/response style of conversation, which makes sense, but is not something obviously guaranteed by the API. I would suggest restructuring it like:

```ts
if (historyItem instanceof vscode.ChatResponseTurn) {
  databaseName = databaseName || historyItem.result.metadata?.databaseName;
  collectionName = collectionName || historyItem.result.metadata?.collectionName;
} else if (historyItem instanceof vscode.ChatRequestTurn) {
  databaseName = databaseName || historyItem.prompt.match(databaseNameInCommandResponseKeyRegex)?.[1];
  collectionName = collectionName || historyItem.prompt.match(collectionNameInCommandResponseKeyRegex)?.[1];
}

if (databaseName && collectionName) {
  // If we've found the entire namespace, stop looping
  break;
}
```

Note that here I'm assuming reverse iteration of the history so short-circuiting evaluation of the components if we've already found a match.